File: src-api/src/app/api/agent.ts
import { Hono } from 'hono';

import {
  createSession,
  deleteSession,
  getSession,
  runAgent,
} from '@/shared/services/agent';
import type { AgentRequest } from '@/shared/types/agent';

const agent = new Hono();

// Helper to create SSE stream
function createSSEStream(generator: AsyncGenerator<unknown>) {
  const encoder = new TextEncoder();
  return new ReadableStream({
    async start(controller) {
      try {
        for await (const message of generator) {
          const data = `data: ${JSON.stringify(message)}\n\n`;
          controller.enqueue(encoder.encode(data));
        }
      } catch (error) {
        const errorData = `data: ${JSON.stringify({
          type: 'error',
          message: error instanceof Error ? error.message : String(error),
        })}\n\n`;
        controller.enqueue(encoder.encode(errorData));
      } finally {
        controller.close();
      }
    },
  });
}

// SSE Response headers
const SSE_HEADERS = {
  'Content-Type': 'text/event-stream',
  'Cache-Control': 'no-cache, no-transform',
  Connection: 'keep-alive',
  'X-Accel-Buffering': 'no',
};

// Direct execution - Claude Agent SDK handles task planning internally
agent.post('/', async (c) => {
  const body = await c.req.json<AgentRequest>();

  console.log('[AgentAPI] POST / received:', {
    hasPrompt: !!body.prompt,
    hasModelConfig: !!body.modelConfig,
    modelConfig: body.modelConfig
      ? {
          hasApiKey: !!body.modelConfig.apiKey,
          baseUrl: body.modelConfig.baseUrl,
          model: body.modelConfig.model,
        }
      : null,
    sandboxConfig: body.sandboxConfig
      ? {
          enabled: body.sandboxConfig.enabled,
          provider: body.sandboxConfig.provider,
        }
      : null,
    hasImages: !!body.images,
    imagesCount: body.images?.length || 0,
  });

  // Debug logging for images
  if (body.images && body.images.length > 0) {
    body.images.forEach(
      (img: { data: string; mimeType: string }, i: number) => {
        console.log(
          `[AgentAPI] Image ${i}: mimeType=${img.mimeType}, dataLength=${img.data?.length || 0}`
        );
      }
    );
  } else {
    console.log('[AgentAPI] No images in request');
  }

  if (!body.prompt) {
    return c.json({ error: 'prompt is required' }, 400);
  }

  const session = createSession();
  const readable = createSSEStream(
    runAgent(
      body.prompt,
      session,
      body.conversation,
      body.workDir,
      body.taskId,
      body.modelConfig,
      body.sandboxConfig,
      body.images,
      body.skillsConfig,
      body.mcpConfig
    )
  );

  return new Response(readable, { headers: SSE_HEADERS });
});

// Stop a running agent
agent.post('/stop/:sessionId', async (c) => {
  const sessionId = c.req.param('sessionId');
  const session = getSession(sessionId);

  if (!session) {
    return c.json({ error: 'Session not found' }, 404);
  }

  deleteSession(sessionId);
  return c.json({ status: 'stopped' });
});

// Get session status
agent.get('/session/:sessionId', async (c) => {
  const sessionId = c.req.param('sessionId');
  const session = getSession(sessionId);

  if (!session) {
    return c.json({ error: 'Session not found' }, 404);
  }

  return c.json({
    id: session.id,
    createdAt: session.createdAt,
    phase: session.phase,
    isAborted: session.abortController.signal.aborted,
  });
});

export default agent;

File: src-python/app/api/agent.py
"""Agent API routes."""
import json
from typing import Optional

from fastapi import APIRouter
from fastapi.responses import StreamingResponse
from pydantic import BaseModel

from shared.services.agent import (
    create_session,
    delete_session,
    get_session,
    run_agent,
)
from shared.types.agent import (
    ConversationMessage,
    ImageAttachment,
    McpConfig,
    SandboxConfig,
    SkillsConfig,
)

router = APIRouter()


class ModelConfig(BaseModel):
    """Model configuration."""
    apiKey: Optional[str] = None
    baseUrl: Optional[str] = None
    model: Optional[str] = None


class AgentRequest(BaseModel):
    """Agent execution request."""
    prompt: str
    conversation: Optional[list[ConversationMessage]] = None
    workDir: Optional[str] = None
    taskId: Optional[str] = None
    modelConfig: Optional[ModelConfig] = None
    sandboxConfig: Optional[SandboxConfig] = None
    images: Optional[list[ImageAttachment]] = None
    skillsConfig: Optional[SkillsConfig] = None
    mcpConfig: Optional[McpConfig] = None


async def create_sse_stream(prompt: str, request: AgentRequest):
    """Create SSE stream for agent execution.
    
    Args:
        prompt: User prompt
        request: Agent request
    
    Yields:
        SSE formatted messages
    """
    session = create_session()
    
    # Convert ModelConfig to dict
    model_config_dict = None
    if request.modelConfig:
        model_config_dict = request.modelConfig.model_dump()
    
    try:
        async for message in run_agent(
            prompt=prompt,
            session=session,
            conversation=request.conversation,
            work_dir=request.workDir,
            task_id=request.taskId,
            model_config=model_config_dict,
            sandbox_config=request.sandboxConfig,
            images=request.images,
            skills_config=request.skillsConfig,
            mcp_config=request.mcpConfig,
        ):
            # Format as SSE
            data = json.dumps(message)
            yield f"data: {data}\n\n"
    except Exception as e:
        # Send error message
        error_data = json.dumps({
            "type": "error",
            "message": str(e),
        })
        yield f"data: {error_data}\n\n"


@router.post("/")
async def execute_agent(request: AgentRequest):
    """Execute agent with the given prompt.
    
    Args:
        request: Agent execution request
    
    Returns:
        SSE stream of agent messages
    """
    return StreamingResponse(
        create_sse_stream(request.prompt, request),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache, no-transform",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        },
    )


@router.post("/stop/{session_id}")
async def stop_agent_execution(session_id: str):
    """Stop a running agent.
    
    Args:
        session_id: Session ID to stop
    
    Returns:
        Status message
    """
    session = get_session(session_id)
    if not session:
        return {"error": "Session not found"}, 404
    
    delete_session(session_id)
    return {"status": "stopped"}


@router.get("/session/{session_id}")
async def get_agent_session(session_id: str):
    """Get session status.
    
    Args:
        session_id: Session ID
    
    Returns:
        Session information
    """
    session = get_session(session_id)
    if not session:
        return {"error": "Session not found"}, 404
    
    return {
        "id": session.id,
        "createdAt": session.created_at.isoformat(),
        "phase": session.phase,
        "isAborted": session.is_aborted,
    }
